wurst/math/Vectors.wurst

```diff
@@ -394,6 +401,13 @@ public function vec3.assertEquals(vec3 expected)
 	if compare.x > delta or compare.y > delta or compare.z > delta
 		testFail("Expected <{0}, Actual <{1} with delta {2}>".format(expected.toString(), this.toString(), delta.toString()))
 
+/** Checks whether the point is in a 2D triangle defined by 3 points. */
+public function vec3.isInTriangle2d(vec3 p1, vec3 p2, vec3 p3) returns bool
```

Comment:
Kind of misleading to use vec3 here, although I understand your intent.
Maybe add an explicit comment that this discards the Z coordinate?

Also, to avoid code duplication, you can instead just implement this as:
`return this.toVec2().isInTriangle(p1.toVec2(), p2.toVec2(), p3.toVec2())`

I think the generated JASS code will be virtually identical, but it might be worth checking anyway.

Reply:
> Kind of misleading to use vec3 here, although I understand your intent.
Maybe add an explicit comment that this discards the Z coordinate?

Yes, it would be better.

> Also, to avoid code duplication, you can instead just implement this as:
return this.toVec2().isInTriangle(p1.toVec2(), p2.toVec2(), p3.toVec2())

I tried. 

<details> <summary>This is the result of compilation for current version</summary>

```
function init_Hello takes nothing returns boolean
	local player p = Player(0)
	local integer unitId = 1751543663
	local real temp_pos_x = 0.
	local real temp_pos_y = 0.
	local real pos_x = temp_pos_x
	local real pos_y = temp_pos_y
	local real facing_radians = 0.
	local player temp = p
	local integer temp_2 = unitId
	local real temp_3 = pos_x
	local real temp_4 = pos_y
	local real this_radians = facing_radians
	local real p_x
	local real p_y
	local real p_z
	local real p1_x
	local real p1_y
	local real p1_z
	local real p2_x
	local real p2_y
	local real p2_z
	local real p3_x
	local real p3_y
	local real p3_z
	local boolean b
	local real this_x
	local real this_y
	local real p1_x_2
	local real p1_y_2
	local real p2_x_2
	local real p2_y_2
	local real p3_x_2
	local real p3_y_2
	local real a
	local real b_2
	local real c
	local real temp_p_x
	local real temp_p_y
	local real temp_p_z
	local real temp_p1_x
	local real temp_p1_y
	local real temp_p1_z
	local real temp_p2_x
	local real temp_p2_y
	local real temp_p2_z
	local real temp_p3_x
	local real temp_p3_y
	local real temp_p3_z
	local real temp_this_x
	local real temp_this_y
	local real temp_p1_x_2
	local real temp_p1_y_2
	local real temp_p2_x_2
	local real temp_p2_y_2
	local real temp_p3_x_2
	local real temp_p3_y_2
	set Hello_u = CreateUnit(temp, temp_2, temp_3, temp_4, this_radians * 57.295779513)
	set temp_p_x = unit_getPos3Real(Hello_u)
	set temp_p_y = unit_getPos3Real_return_y
	set temp_p_z = unit_getPos3Real_return_z
	set p_x = temp_p_x
	set p_y = temp_p_y
	set p_z = temp_p_z
	set temp_p1_x = -150.
	set temp_p1_y = -250.
	set temp_p1_z = 0.
	set p1_x = temp_p1_x
	set p1_y = temp_p1_y
	set p1_z = temp_p1_z
	set temp_p2_x = 343.
	set temp_p2_y = 156.6
	set temp_p2_z = 0.
	set p2_x = temp_p2_x
	set p2_y = temp_p2_y
	set p2_z = temp_p2_z
	set temp_p3_x = -235.
	set temp_p3_y = 335.
	set temp_p3_z = 0.
	set p3_x = temp_p3_x
	set p3_y = temp_p3_y
	set p3_z = temp_p3_z
	set temp_this_x = p_x
	set temp_this_y = p_y
	set this_x = temp_this_x
	set this_y = temp_this_y
	set temp_p1_x_2 = p1_x
	set temp_p1_y_2 = p1_y
	set p1_x_2 = temp_p1_x_2
	set p1_y_2 = temp_p1_y_2
	set temp_p2_x_2 = p2_x
	set temp_p2_y_2 = p2_y
	set p2_x_2 = temp_p2_x_2
	set p2_y_2 = temp_p2_y_2
	set temp_p3_x_2 = p3_x
	set temp_p3_y_2 = p3_y
	set p3_x_2 = temp_p3_x_2
	set p3_y_2 = temp_p3_y_2
	set a = this_x * (p1_y_2 - p2_y_2) + this_y * (p2_x_2 - p1_x_2) + (p1_x_2 * p2_y_2 - p2_x_2 * p1_y_2)
	set b_2 = this_x * (p2_y_2 - p3_y_2) + this_y * (p3_x_2 - p2_x_2) + (p2_x_2 * p3_y_2 - p3_x_2 * p2_y_2)
	set c = this_x * (p3_y_2 - p1_y_2) + this_y * (p1_x_2 - p3_x_2) + (p3_x_2 * p1_y_2 - p1_x_2 * p3_y_2)
	set b = (a <= 0. and b_2 <= 0. and c <= 0.) or (a >= 0. and b_2 >= 0. and c >= 0.)
	if b then
	endif
	set p = null
	set temp = null
	return true
endfunction
```
</details>

<details> <summary>And through .toVec2</summary>

```
function vec3_isInTriangle2d takes real this_x, real this_y, real this_z, real p1_x, real p1_y, real p1_z, real p2_x, real p2_y, real p2_z, real p3_x, real p3_y, real p3_z returns boolean
	local real temp_this_x = this_x
	local real temp_this_y = this_y
	local real this_x_3 = temp_this_x
	local real this_y_3 = temp_this_y
	local real temp_this_x_2 = this_x_3
	local real temp_this_y_2 = this_y_3
	local real this_x_2 = temp_this_x_2
	local real this_y_2 = temp_this_y_2
	local real temp_this_x_3 = p1_x
	local real temp_this_y_3 = p1_y
	local real this_x_4 = temp_this_x_3
	local real this_y_4 = temp_this_y_3
	local real temp_p1_x = this_x_4
	local real temp_p1_y = this_y_4
	local real p1_x_2 = temp_p1_x
	local real p1_y_2 = temp_p1_y
	local real temp_this_x_4 = p2_x
	local real temp_this_y_4 = p2_y
	local real this_x_5 = temp_this_x_4
	local real this_y_5 = temp_this_y_4
	local real temp_p2_x = this_x_5
	local real temp_p2_y = this_y_5
	local real p2_x_2 = temp_p2_x
	local real p2_y_2 = temp_p2_y
	local real temp_this_x_5 = p3_x
	local real temp_this_y_5 = p3_y
	local real this_x_6 = temp_this_x_5
	local real this_y_6 = temp_this_y_5
	local real temp_p3_x = this_x_6
	local real temp_p3_y = this_y_6
	local real p3_x_2 = temp_p3_x
	local real p3_y_2 = temp_p3_y
	local real a = this_x_2 * (p1_y_2 - p2_y_2) + this_y_2 * (p2_x_2 - p1_x_2) + (p1_x_2 * p2_y_2 - p2_x_2 * p1_y_2)
	local real b = this_x_2 * (p2_y_2 - p3_y_2) + this_y_2 * (p3_x_2 - p2_x_2) + (p2_x_2 * p3_y_2 - p3_x_2 * p2_y_2)
	local real c = this_x_2 * (p3_y_2 - p1_y_2) + this_y_2 * (p1_x_2 - p3_x_2) + (p3_x_2 * p1_y_2 - p1_x_2 * p3_y_2)
	return (a <= 0. and b <= 0. and c <= 0.) or (a >= 0. and b >= 0. and c >= 0.)
endfunction

function init_Hello takes nothing returns boolean
	local player p = Player(0)
	local integer unitId = 1751543663
	local real temp_pos_x = 0.
	local real temp_pos_y = 0.
	local real pos_x = temp_pos_x
	local real pos_y = temp_pos_y
	local real facing_radians = 0.
	local player temp = p
	local integer temp_2 = unitId
	local real temp_3 = pos_x
	local real temp_4 = pos_y
	local real this_radians = facing_radians
	local real p_x
	local real p_y
	local real p_z
	local real p1_x
	local real p1_y
	local real p1_z
	local real p2_x
	local real p2_y
	local real p2_z
	local real p3_x
	local real p3_y
	local real p3_z
	local boolean b
	local real temp_p_x
	local real temp_p_y
	local real temp_p_z
	local real temp_p1_x
	local real temp_p1_y
	local real temp_p1_z
	local real temp_p2_x
	local real temp_p2_y
	local real temp_p2_z
	local real temp_p3_x
	local real temp_p3_y
	local real temp_p3_z
	set Hello_u = CreateUnit(temp, temp_2, temp_3, temp_4, this_radians * 57.295779513)
	set temp_p_x = unit_getPos3Real(Hello_u)
	set temp_p_y = unit_getPos3Real_return_y
	set temp_p_z = unit_getPos3Real_return_z
	set p_x = temp_p_x
	set p_y = temp_p_y
	set p_z = temp_p_z
	set temp_p1_x = -150.
	set temp_p1_y = -250.
	set temp_p1_z = 0.
	set p1_x = temp_p1_x
	set p1_y = temp_p1_y
	set p1_z = temp_p1_z
	set temp_p2_x = 343.
	set temp_p2_y = 156.6
	set temp_p2_z = 0.
	set p2_x = temp_p2_x
	set p2_y = temp_p2_y
	set p2_z = temp_p2_z
	set temp_p3_x = -235.
	set temp_p3_y = 335.
	set temp_p3_z = 0.
	set p3_x = temp_p3_x
	set p3_y = temp_p3_y
	set p3_z = temp_p3_z
	set b = vec3_isInTriangle2d(p_x, p_y, p_z, p1_x, p1_y, p1_z, p2_x, p2_y, p2_z, p3_x, p3_y, p3_z)
	if b then
	endif
	set p = null
	set temp = null
	return true
endfunction
```
</details>

We've got more local declarations and one more function call in case `.toVec2`.

Reply:
You do realize though, that comparing unotpmized code is completely pointless, right?
`.toVec2()` calls simply drop the z component. This function call will be optimized away completely, pretty much always.

This
```
let u = createUnit(Player(0), 'hfoo', vec2(0, 0), angle(0))

init
	let p = u.getPos3Real()
	let p1 = vec3(-150, -250, 0)
	let p2 = vec3(343, 156.6, 0)
	let p3 = vec3(-235, 335, 0)
	let b = p.isInTriangle2d(p1, p2, p3)
	print(b)
```

turns into:
```
function init_Test takes nothing returns boolean
	local real a
	local real b
	local string msgString
	local real temp_p_x
	local real temp_p_y
	set Projectile_u = CreateUnit(Player(0), 1751543663, 0., 0., 0)
	set temp_p_x = unit_getPos3Real(Projectile_u)
	set temp_p_y = unit_getPos3Real_return_y
	set a = temp_p_x * -406.6 + temp_p_y * 493 + 62260
	set b = temp_p_x * -178.4 + temp_p_y * -578 + 151706
	set temp_p_y = temp_p_x * 585 + temp_p_y * 85 + 109000
	set msgString = "False"
	if (a <= 0. and b <= 0. and temp_p_y <= 0.) or (a >= 0. and b >= 0. and temp_p_y >= 0.) then
		set msgString = "True"
	endif
	call DisplayTimedTextToPlayer(Player_localPlayer, 0., 0., 45., msgString)
	return true
endfunction
```

For **BOTH** versions. The only issue I had was that the vec3 variant wasn't getting inlined by the heuristic (used @inline for the example).

Reply:
Just tell me what you want. Was it a requirement or a suggestion.

Reply:
Use the 1 liner and add `@inline`